### PR TITLE
planner: fix wrong collation when rewrite in condition

### LINKED ISSUE #30486: Collation used by `IN` clause is not correct.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

``` sql
create table t(a char(10), b char(10));
insert into t values ("a", "a");
insert into t values ("A", "A");
select * from t where a in ('b' collate utf8mb4_general_ci, 'A', 3);
```

### 2. What did you expect to see? (Required)
```
mysql> select * from t where a in ('b' collate utf8mb4_general_ci, 'A', 3);
+------+------+
| a    | b    |
+------+------+
| a    | a    |
| A    | A    |
+------+------+
2 rows in set (0.00 sec)
```
### 3. What did you see instead (Required)
```
mysql> select * from t where a in ('b' collate utf8mb4_general_ci, 'A', 3);
+------+------+
| a    | b    |
+------+------+
| A    | A    |
+------+------+
1 row in set, 1 warning (0.01 sec)
```
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #30486

Problem Summary:
When rewriting IN to EQ, we need to set the collation correctly.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by bb7133 — APPROVED** [review]
LGTM

**Review by xhebox — APPROVED** [review]

**Review by xhebox — CHANGES_REQUESTED** [review]
Then failed tests look stable, @wjhuang2016 :

```
prepare stmt from 'select col1 from t where (col1, col2) in ((?, 1));';, [], error stack [expression:1241]Operand should contain 2 column(s)
```

File: planner/core/expression_rewriter.go [hunk 1]
```diff
@@ -1481,8 +1481,19 @@ func (er *expressionRewriter) inToExpression(lLen int, not bool, tp *types.Field
...
 		function = er.notToExpression(not, ast.In, tp, er.ctxStack[stkLen-lLen-1:]...)
 	} else {
+		coll, err := expression.CheckAndDeriveCollationFromExprs(er.sctx, "IN", types.ETInt, args...)
+		er.err = err
+		if er.err != nil {
+			return
+		}
 		eqFunctions := make([]expression.Expression, 0, lLen)
 		for i := stkLen - lLen; i < stkLen; i++ {
+			if er.ctxStack[i].GetType().EvalType() == types.ETString && l == 1 {
```
Comment (bb7133) [inline planner/core/expression_rewriter.go:1491]: Well, I got confused by the condition `l == 1`, could you explain it and add some comments here?

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -283,6 +283,8 @@ func (c *castAsStringFunctionClass) getFunction(ctx sessionctx.Context, args []E
 	if err != nil {
 		return nil, err
 	}
+
+	bf.args[0] = HandleBinaryLiteral(ctx, args[0], &ExprCollation{Charset: c.tp.Charset, Collation: c.tp.Collate}, c.funcName)
```
Comment (wjhuang2016) [inline expression/builtin_cast.go:287]: Wait for another PR to fix it.

File: expression/builtin_convert_charset.go [hunk 3]
```diff
@@ -263,7 +263,7 @@ func HandleBinaryLiteral(ctx sessionctx.Context, expr Expression, ec *ExprCollat
 		ast.Left, ast.Right, ast.Repeat, ast.Trim, ast.LTrim, ast.RTrim, ast.Substr, ast.SubstringIndex, ast.Replace,
 		ast.Substring, ast.Mid, ast.Translate, ast.InsertFunc, ast.Lpad, ast.Rpad, ast.Elt, ast.ExportSet, ast.MakeSet,
 		ast.FindInSet, ast.Regexp, ast.Field, ast.Locate, ast.Instr, ast.Position, ast.GE, ast.LE, ast.GT, ast.LT, ast.EQ,
-		ast.NE, ast.NullEQ, ast.Strcmp, ast.If, ast.Ifnull, ast.Like, ast.In, ast.DateFormat, ast.TimeFormat:
+		ast.NE, ast.NullEQ, ast.Strcmp, ast.If, ast.Ifnull, ast.Like, ast.In, ast.DateFormat, ast.TimeFormat, ast.Cast:
```
Comment (wjhuang2016) [inline expression/builtin_convert_charset.go:266]: Wait for another PR to fix it.

File: expression/builtin_cast.go (hunk 2 above)
Comment (bb7133) [reply expression/builtin_cast.go:287]: Which PR is it?

**Review by xiongjiwei — APPROVED** [review]

**Review by xhebox — APPROVED** [review]

**Pull request merged**